File: apps/ios/GalaxySSI/AgentAttachmentRecoveryProtocol.swift
import Foundation

struct AgentAttachmentRecoveryRequest: Equatable {
  static let requestType = "input_attachment_request"
  static let resultType = "input_attachment_request_result"

  var requestId: String
  var clientRouteId: String
  var conversationId: String
  var taskId: String
  var turnId: String
  var contactId: String
  var sourceMessageId: Int64
  var attachmentIds: [String]

  static func decode(_ payload: AgentMcpJSONObject) -> AgentAttachmentRecoveryRequest? {
    guard payload.string("type") == requestType else {
      return nil
    }
    let requestId = payload.string("request_id").trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    guard isValidRequestId(requestId),
          let sourceMessageId = sourceMessageId(from: payload["source_message_id"]),
          sourceMessageId > 0,
          let clientRouteId = identity(payload, "client_route_id"),
          let conversationId = identity(payload, "conversation_id"),
          let taskId = identity(payload, "task_id"),
          let turnId = identity(payload, "turn_id"),
          let contactId = identity(payload, "contact_id") else {
      return nil
    }
    let attachmentIds = requestedAttachmentIds(payload["attachment_ids"]?.arrayValue ?? [])
    guard !attachmentIds.isEmpty else {
      return nil
    }
    return AgentAttachmentRecoveryRequest(
      requestId: requestId,
      clientRouteId: clientRouteId,
      conversationId: conversationId,
      taskId: taskId,
      turnId: turnId,
      contactId: contactId,
      sourceMessageId: sourceMessageId,
      attachmentIds: attachmentIds
    )
  }

  func result(
    status: String,
    availableAttachmentIds: [String] = [],
    missingAttachmentIds: [String] = [],
    error: String = "",
    timeMillis: Int64 = AgentAttachmentRecoveryRequest.currentTimeMillis()
  ) -> AgentMcpJSONObject {
    var payload: AgentMcpJSONObject = [
      "type": .string(Self.resultType),
      "request_id": .string(requestId),
      "client_route_id": .string(clientRouteId),
      "conversation_id": .string(conversationId),
      "task_id": .string(taskId),
      "turn_id": .string(turnId),
      "contact_id": .string(contactId),
      "source_message_id": .string(String(sourceMessageId)),
      "status": .string(status),
      "available_attachment_ids": .array(availableAttachmentIds.map { .string($0) }),
      "missing_attachment_ids": .array(missingAttachmentIds.map { .string($0) }),
      "time": .int(timeMillis)
    ]
    let trimmedError = error.trimmingCharacters(in: .whitespacesAndNewlines)
    if !trimmedError.isEmpty {
      payload["error"] = .string(String(trimmedError.prefix(Self.maxErrorCharacters)))
    }
    return payload
  }

  private static func requestedAttachmentIds(_ values: [AgentMcpJSONValue]) -> [String] {
    var result: [String] = []
    for value in values.prefix(maxAttachments) {
      let attachmentId = (value.stringValue ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
      guard !attachmentId.isEmpty,
            attachmentId.count <= maxAttachmentIdCharacters,
            !result.contains(attachmentId) else {
        continue
      }
      result.append(attachmentId)
    }
    return result
  }

  private static func identity(_ payload: AgentMcpJSONObject, _ key: String) -> String? {
    let value = payload.string(key).trimmingCharacters(in: .whitespacesAndNewlines)
    guard !value.isEmpty,
          value.count <= maxIdentityCharacters,
          !value.unicodeScalars.contains(where: { $0.value < 32 }) else {
      return nil
    }
    return value
  }

  private static func sourceMessageId(from value: AgentMcpJSONValue?) -> Int64? {
    switch value {
    case .string(let raw):
      return Int64(raw.trimmingCharacters(in: .whitespacesAndNewlines))
    case .int(let raw):
      return raw
    case .double(let raw) where raw.isFinite && raw.rounded(.towardZero) == raw:
      return Int64(raw)
    default:
      return nil
    }
  }

  private static func isValidRequestId(_ value: String) -> Bool {
    value.count == 32 && value.unicodeScalars.allSatisfy { scalar in
      let value = scalar.value
      return (value >= 48 && value <= 57) || (value >= 97 && value <= 102)
    }
  }

  private static func currentTimeMillis() -> Int64 {
    Int64(Date().timeIntervalSince1970 * 1_000)
  }

  private static let maxAttachments = 10
  private static let maxAttachmentIdCharacters = 120
  private static let maxIdentityCharacters = 256
  private static let maxErrorCharacters = 300
}

File: apps/android/app/src/main/java/com/galaxyssi/chat/AgentAttachmentRecoveryProtocol.kt
package com.galaxyssi.chat

import org.json.JSONArray
import org.json.JSONObject

internal data class AgentAttachmentRecoveryRequest(
    val requestId: String,
    val clientRouteId: String,
    val conversationId: String,
    val taskId: String,
    val turnId: String,
    val contactId: String,
    val sourceMessageId: Long,
    val attachmentIds: List<String>
) {
    fun result(
        status: String,
        availableAttachmentIds: Collection<String> = emptyList(),
        missingAttachmentIds: Collection<String> = emptyList(),
        error: String = ""
    ): JSONObject = JSONObject()
        .put("type", RESULT_TYPE)
        .put("request_id", requestId)
        .put("client_route_id", clientRouteId)
        .put("conversation_id", conversationId)
        .put("task_id", taskId)
        .put("turn_id", turnId)
        .put("contact_id", contactId)
        .put("source_message_id", sourceMessageId.toString())
        .put("status", status)
        .put("available_attachment_ids", JSONArray(availableAttachmentIds.toList()))
        .put("missing_attachment_ids", JSONArray(missingAttachmentIds.toList()))
        .put("time", System.currentTimeMillis())
        .also { payload ->
            error.trim().takeIf(String::isNotBlank)?.let {
                payload.put("error", it.take(300))
            }
        }

    companion object {
        const val REQUEST_TYPE = "input_attachment_request"
        const val RESULT_TYPE = "input_attachment_request_result"
        private const val MAX_ATTACHMENTS = 10
        private val REQUEST_ID = Regex("[a-f0-9]{32}")

        fun decode(payload: JSONObject): AgentAttachmentRecoveryRequest? = runCatching {
            require(payload.optString("type") == REQUEST_TYPE)
            val requestId = payload.getString("request_id").lowercase()
            require(requestId.matches(REQUEST_ID))
            val sourceMessageId = payload.optString("source_message_id").toLongOrNull()
                ?: payload.optLong("source_message_id", 0L)
            require(sourceMessageId > 0L)
            val attachmentIds = payload.optJSONArray("attachment_ids") ?: JSONArray()
            val requested = buildList {
                for (index in 0 until minOf(attachmentIds.length(), MAX_ATTACHMENTS)) {
                    val value = attachmentIds.optString(index).trim()
                    if (value.isNotBlank() && value.length <= 120 && value !in this) add(value)
                }
            }
            require(requested.isNotEmpty())
            AgentAttachmentRecoveryRequest(
                requestId = requestId,
                clientRouteId = identity(payload, "client_route_id"),
                conversationId = identity(payload, "conversation_id"),
                taskId = identity(payload, "task_id"),
                turnId = identity(payload, "turn_id"),
                contactId = identity(payload, "contact_id"),
                sourceMessageId = sourceMessageId,
                attachmentIds = requested
            )
        }.getOrNull()

        private fun identity(payload: JSONObject, key: String): String {
            val value = payload.getString(key).trim()
            require(value.isNotBlank() && value.length <= 256 && value.none { it.code < 32 })
            return value
        }
    }
}
